File: services/event-streaming-service/src/runtime/flink/deployer.rs
//! Apache Flink Kubernetes Operator deployer.
//!
//! Materialises a topology as:
//!
//!   1. A `ConfigMap/{deployment}-sql` with the rendered Flink SQL.
//!   2. A `FlinkDeployment/{deployment}` running the `sql-runner.jar`
//!      image, with `args` pointing at the ConfigMap.
//!   3. A `FlinkSessionJob` is intentionally skipped — `mode: native`
//!      runs the job inside the same FlinkDeployment, which is the
//!      shape used by `infra/k8s/flink/flinkdeployment-cdc-iceberg.yaml`
//!      in production. The `FlinkSessionJob` path is left as a TODO.
//!
//! Both resources are upserted via SSA (`Patch::Apply`) so the
//! deployer is idempotent.

use kube::api::{Api, ApiResource, DynamicObject, Patch, PatchParams, PostParams};
use kube::Client;
use serde_json::{json, Value};
use sqlx::PgPool;
use uuid::Uuid;

use crate::models::stream::StreamDefinition;
use crate::models::topology::TopologyDefinition;
use crate::runtime::flink::sql::{render_flink_sql, RenderedFlinkSql};
use crate::runtime::flink::{FlinkJobCoords, FlinkRuntimeConfig};

const FIELD_MANAGER: &str = "event-streaming-service";

/// Outcome of a single deployment.
#[derive(Debug, Clone)]
pub struct DeploymentReport {
    pub coords: FlinkJobCoords,
    pub sql: RenderedFlinkSql,
}

/// Errors surfaced by the deployer. We collapse the wide
/// `kube::Error`/`sqlx::Error` graphs into strings at the boundary so the
/// REST handlers can map them to HTTP 5xx with a useful message.
#[derive(Debug, thiserror::Error)]
pub enum DeployerError {
    #[error("kube client: {0}")]
    Kube(String),
    #[error("database: {0}")]
    Db(#[from] sqlx::Error),
}

/// Deploy `topology` onto the Flink cluster reachable via the in-cluster
/// kube API. Updates `streaming_topologies.flink_deployment_name`
/// and `flink_namespace` on success.
pub async fn deploy_topology(
    db: &PgPool,
    cfg: &FlinkRuntimeConfig,
    topology: &TopologyDefinition,
    streams: &[StreamDefinition],
) -> Result<DeploymentReport, DeployerError> {
    let namespace = topology
        .flink_namespace
        .clone()
        .unwrap_or_else(|| cfg.default_namespace.clone());
    let deployment_name = topology
        .flink_deployment_name
        .clone()
        .or_else(|| topology.flink_job_name.clone())
        .unwrap_or_else(|| format!("topo-{}", topology.id.simple()));

    let sql = render_flink_sql(topology, streams);

    let client = Client::try_default()
        .await
        .map_err(|e| DeployerError::Kube(e.to_string()))?;

    apply_sql_configmap(&client, &namespace, &deployment_name, &sql.script).await?;
    apply_flink_deployment(&client, cfg, &namespace, &deployment_name, topology).await?;

    sqlx::query(
        "UPDATE streaming_topologies
            SET flink_deployment_name = $2,
                flink_namespace = $3,
                runtime_kind = 'flink',
                updated_at = now()
          WHERE id = $1",
    )
    .bind(topology.id)
    .bind(&deployment_name)
    .bind(&namespace)
    .execute(db)
    .await?;

    Ok(DeploymentReport {
        coords: FlinkJobCoords {
            deployment_name,
            namespace,
            job_id: None,
        },
        sql,
    })
}

async fn apply_sql_configmap(
    client: &Client,
    namespace: &str,
    deployment: &str,
    script: &str,
) -> Result<(), DeployerError> {
    let api: Api<k8s_openapi::api::core::v1::ConfigMap> =
        Api::namespaced(client.clone(), namespace);
    let cm_name = format!("{deployment}-sql");
    let cm = k8s_openapi::api::core::v1::ConfigMap {
        metadata: k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta {
            name: Some(cm_name.clone()),
            namespace: Some(namespace.to_string()),
            labels: Some(
                [
                    ("app.kubernetes.io/managed-by".to_string(), FIELD_MANAGER.to_string()),
                    ("openfoundry.io/component".to_string(), "flink-sql".to_string()),
                ]
                .into_iter()
                .collect(),
            ),
            ..Default::default()
        },
        data: Some(
            [("topology.sql".to_string(), script.to_string())]
                .into_iter()
                .collect(),
        ),
        ..Default::default()
    };
    api.patch(
        &cm_name,
        &PatchParams::apply(FIELD_MANAGER).force(),
        &Patch::Apply(&cm),
    )
    .await
    .map_err(|e| DeployerError::Kube(format!("apply ConfigMap/{cm_name}: {e}")))?;
    Ok(())
}

async fn apply_flink_deployment(
    client: &Client,
    cfg: &FlinkRuntimeConfig,
    namespace: &str,
    deployment: &str,
    topology: &TopologyDefinition,
) -> Result<(), DeployerError> {
    let resource = ApiResource {
        group: "flink.apache.org".into(),
        version: "v1beta1".into(),
        api_version: "flink.apache.org/v1beta1".into(),
        kind: "FlinkDeployment".into(),
        plural: "flinkdeployments".into(),
    };
    let api: Api<DynamicObject> = Api::namespaced_with(client.clone(), namespace, &resource);

    let body = render_flink_deployment_manifest(cfg, namespace, deployment, topology);
    // We round-trip through DynamicObject so kube's typed Patch::Apply
    // gets the right TypeMeta wired up.
    let dyn_obj: DynamicObject = serde_json::from_value(body)
        .map_err(|e| DeployerError::Kube(format!("manifest serialise: {e}")))?;

    match api
        .patch(
            deployment,
            &PatchParams::apply(FIELD_MANAGER).force(),
            &Patch::Apply(&dyn_obj),
        )
        .await
    {
        Ok(_) => Ok(()),
        Err(kube::Error::Api(err)) if err.code == 404 => {
            // Some operator builds reject SSA on first creation; fall
            // back to a plain create.
            api.create(&PostParams::default(), &dyn_obj)
                .await
                .map_err(|e| DeployerError::Kube(format!("create FlinkDeployment: {e}")))?;
            Ok(())
        }
        Err(e) => Err(DeployerError::Kube(format!(
            "apply FlinkDeployment/{deployment}: {e}"
        ))),
    }
}

/// Pure helper: build the FlinkDeployment manifest as JSON. Exposed for
/// unit tests and for the `--render-only` debug endpoint.
pub fn render_flink_deployment_manifest(
    cfg: &FlinkRuntimeConfig,
    namespace: &str,
    deployment: &str,
    topology: &TopologyDefinition,
) -> Value {
    let checkpoint_dir = format!("{}/checkpoints/{}", cfg.state_bucket_uri, deployment);
    let savepoint_dir = format!("{}/savepoints/{}", cfg.state_bucket_uri, deployment);
    let ha_dir = format!("{}/ha/{}", cfg.state_bucket_uri, deployment);
    let checkpointing_mode = match topology.consistency_guarantee.as_str() {
        "exactly-once" => "EXACTLY_ONCE",
        _ => "AT_LEAST_ONCE",
    };

    json!({
        "apiVersion": "flink.apache.org/v1beta1",
        "kind": "FlinkDeployment",
        "metadata": {
            "name": deployment,
            "namespace": namespace,
            "labels": {
                "app.kubernetes.io/managed-by": FIELD_MANAGER,
                "openfoundry.io/topology-id": topology.id.to_string(),
                "openfoundry.io/topology-name": sanitize_label(&topology.name),
            }
        },
        "spec": {
            "image": cfg.sql_runner_image,
            "flinkVersion": cfg.flink_version,
            "mode": "native",
            "serviceAccount": "flink",
            "flinkConfiguration": {
                "taskmanager.numberOfTaskSlots": "2",
                "parallelism.default": "4",
                "high-availability.type": "kubernetes",
                "high-availability.storageDir": ha_dir,
                "state.backend.type": "rocksdb",
                "state.backend.incremental": "true",
                "state.checkpoints.dir": checkpoint_dir,
                "state.savepoints.dir": savepoint_dir,
                "execution.checkpointing.mode": checkpointing_mode,
                "execution.checkpointing.interval": format!("{}", topology.checkpoint_interval_ms),
                "execution.checkpointing.timeout": "600000",
                "metrics.reporter.prom.factory.class":
                    "org.apache.flink.metrics.prometheus.PrometheusReporterFactory",
                "metrics.reporter.prom.port": "9249",
            },
            "jobManager": {
                "replicas": 1,
                "resource": { "memory": "1024m", "cpu": 1 }
            },
            "taskManager": {
                "resource": { "memory": "2048m", "cpu": 1 }
            },
            "podTemplate": {
                "spec": {
                    "containers": [{
                        "name": "flink-main-container",
                        "volumeMounts": [{
                            "name": "topology-sql",
                            "mountPath": "/opt/flink/usrlib/sql",
                        }]
                    }],
                    "volumes": [{
                        "name": "topology-sql",
                        "configMap": { "name": format!("{deployment}-sql") }
                    }]
                }
            },
            "job": {
                "jarURI": "local:///opt/flink/usrlib/sql-runner.jar",
                "args": ["--script", "/opt/flink/usrlib/sql/topology.sql"],
                "parallelism": 2,
                "upgradeMode": "savepoint",
                "state": "running",
            }
        }
    })
}

fn sanitize_label(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for ch in s.chars().take(63) {
        if ch.is_ascii_alphanumeric() || ch == '-' || ch == '_' || ch == '.' {
            out.push(ch);
        } else {
            out.push('-');
        }
    }
    out
}

/// Tear down the resources owned by the topology. Idempotent.
pub async fn delete_topology(
    cfg: &FlinkRuntimeConfig,
    coords: &FlinkJobCoords,
) -> Result<(), DeployerError> {
    let client = Client::try_default()
        .await
        .map_err(|e| DeployerError::Kube(e.to_string()))?;
    let resource = ApiResource {
        group: "flink.apache.org".into(),
        version: "v1beta1".into(),
        api_version: "flink.apache.org/v1beta1".into(),
        kind: "FlinkDeployment".into(),
        plural: "flinkdeployments".into(),
    };
    let api: Api<DynamicObject> =
        Api::namespaced_with(client.clone(), &coords.namespace, &resource);
    let _ = api
        .delete(&coords.deployment_name, &Default::default())
        .await
        .map_err(|e| {
            tracing::warn!("delete FlinkDeployment failed (ignored): {e}");
            DeployerError::Kube(e.to_string())
        });
    let cm: Api<k8s_openapi::api::core::v1::ConfigMap> =
        Api::namespaced(client, &coords.namespace);
    let _ = cm
        .delete(
            &format!("{}-sql", coords.deployment_name),
            &Default::default(),
        )
        .await;
    let _ = cfg; // future: clean up state bucket prefixes
    let _: Option<Uuid> = None; // silence unused import lint
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    #[test]
    fn manifest_carries_topology_metadata_and_checkpointing() {
        let cfg = FlinkRuntimeConfig {
            default_namespace: "flink".into(),
            sql_runner_image: "ghcr.io/x/sql-runner:1.19".into(),
            flink_version: "v1_19".into(),
            jobmanager_url_template: "http://{deployment}-rest.{namespace}.svc:8081".into(),
            metrics_poll_interval_ms: 15_000,
            state_bucket_uri: "s3://bucket/flink".into(),
        };
        let topology = sample_topology();
        let manifest =
            render_flink_deployment_manifest(&cfg, "flink", "topo-demo", &topology);
        assert_eq!(manifest["kind"], Value::String("FlinkDeployment".into()));
        assert_eq!(
            manifest["metadata"]["labels"]["openfoundry.io/topology-id"],
            Value::String(topology.id.to_string())
        );
        assert_eq!(
            manifest["spec"]["flinkConfiguration"]["execution.checkpointing.mode"],
            Value::String("EXACTLY_ONCE".into())
        );
        assert_eq!(
            manifest["spec"]["flinkConfiguration"]["state.checkpoints.dir"],
            Value::String("s3://bucket/flink/checkpoints/topo-demo".into())
        );
    }

    fn sample_topology() -> TopologyDefinition {
        use chrono::Utc;
        TopologyDefinition {
            id: Uuid::now_v7(),
            name: "demo".into(),
            description: String::new(),
            status: "active".into(),
            nodes: Vec::new(),
            edges: Vec::new(),
            join_definition: None,
            cep_definition: None,
            backpressure_policy: crate::models::topology::BackpressurePolicy::default(),
            source_stream_ids: Vec::new(),
            sink_bindings: Vec::new(),
            state_backend: "rocksdb".into(),
            checkpoint_interval_ms: 60_000,
            runtime_kind: "flink".into(),
            flink_job_name: None,
            flink_deployment_name: None,
            flink_job_id: None,
            flink_namespace: None,
            consistency_guarantee: "exactly-once".into(),
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }
}

File: services/ingestion-replication-service/internal/runtime/flink/deployer.go
package flink

// deployer ports event_streaming::runtime::flink::deployer.
//
// Materialises a topology as:
//
//  1. A ConfigMap/{deployment}-sql containing the rendered Flink SQL.
//  2. A FlinkDeployment/{deployment} running the sql-runner.jar image,
//     with args pointing at the ConfigMap.
//
// Both resources are upserted via server-side apply so the deployer is
// idempotent. The Rust source uses kube-rs directly; we keep the Go
// service free of the Kubernetes client dependency by abstracting the
// kube call behind a KubeApplier interface — same pattern HTTPApplier
// in internal/reconcile already uses for the ingest control plane.

import (
	"context"
	"fmt"
	"strings"

	"github.com/openfoundry/openfoundry-go/services/ingestion-replication-service/internal/domain"
)

// FieldManager mirrors deployer::FIELD_MANAGER.
const FieldManager = "event-streaming-service"

// DeploymentReport mirrors deployer::DeploymentReport. It carries the
// coordinates the caller persists back into streaming_topologies plus
// the rendered SQL the operator can preview.
type DeploymentReport struct {
	Coords FlinkJobCoords
	SQL    RenderedFlinkSQL
}

// DeployerErrorKind mirrors the Rust DeployerError thiserror enum.
type DeployerErrorKind int

const (
	DeployerErrUnknown DeployerErrorKind = iota
	DeployerErrKube
	DeployerErrDB
)

// DeployerError mirrors event_streaming::runtime::flink::deployer::DeployerError.
type DeployerError struct {
	Kind    DeployerErrorKind
	Message string
	Cause   error
}

func (e *DeployerError) Error() string {
	switch e.Kind {
	case DeployerErrKube:
		return "kube client: " + e.Message
	case DeployerErrDB:
		if e.Cause != nil {
			return fmt.Sprintf("database: %v", e.Cause)
		}
		return "database: " + e.Message
	default:
		if e.Message != "" {
			return e.Message
		}
		return "unknown deployer error"
	}
}

func (e *DeployerError) Unwrap() error { return e.Cause }

// KubeApplier abstracts the kube-rs server-side apply call. The
// production binding (when added) wraps a REST shim like the existing
// reconcile.HTTPApplier; tests inject a fake.
//
// Apply must be idempotent — multiple Apply calls with the same body
// must converge to the same cluster state. Delete is best-effort: a
// 404/missing-resource must NOT surface as an error.
type KubeApplier interface {
	Apply(ctx context.Context, namespace, kind, name string, manifest map[string]any) error
	Delete(ctx context.Context, namespace, kind, name string) error
}

// CoordsRecorder persists the deployment coordinates back into
// streaming_topologies. Mirrors the
//
//	UPDATE streaming_topologies
//	   SET flink_deployment_name = $2,
//	       flink_namespace       = $3,
//	       runtime_kind          = 'flink',
//	       updated_at            = now()
//	 WHERE id = $1
//
// statement the Rust source runs at the end of deploy_topology.
type CoordsRecorder interface {
	RecordTopologyDeployment(ctx context.Context, topologyID, deploymentName, namespace string) error
}

// Deployer wires the kube applier + DB recorder behind the public
// DeployTopology / DeleteTopology entry points that mirror the Rust
// helpers.
type Deployer struct {
	Applier  KubeApplier
	Recorder CoordsRecorder
}

// DeployTopology mirrors deployer::deploy_topology. It renders the
// topology to Flink SQL, applies the ConfigMap + FlinkDeployment, and
// records the coordinates back into streaming_topologies.
func (d *Deployer) DeployTopology(ctx context.Context, cfg FlinkRuntimeConfig, topology *domain.TopologyDefinition, streams []domain.DomainStreamDefinition) (DeploymentReport, error) {
	if d == nil || d.Applier == nil {
		return DeploymentReport{}, &DeployerError{Kind: DeployerErrKube, Message: "applier is not wired"}
	}
	if topology == nil {
		return DeploymentReport{}, &DeployerError{Kind: DeployerErrKube, Message: "topology is nil"}
	}
	namespace := cfg.DefaultNamespace
	if topology.FlinkNamespace != nil && *topology.FlinkNamespace != "" {
		namespace = *topology.FlinkNamespace
	}
	deployment := ""
	switch {
	case topology.FlinkDeploymentName != nil && *topology.FlinkDeploymentName != "":
		deployment = *topology.FlinkDeploymentName
	case topology.FlinkJobName != nil && *topology.FlinkJobName != "":
		deployment = *topology.FlinkJobName
	default:
		deployment = "topo-" + uuidSimple(topology.ID)
	}

	sql := RenderFlinkSQL(topology, streams)

	cm := RenderSQLConfigMap(namespace, deployment, sql.Script)
	if err := d.Applier.Apply(ctx, namespace, "ConfigMap", deployment+"-sql", cm); err != nil {
		return DeploymentReport{}, &DeployerError{Kind: DeployerErrKube, Message: fmt.Sprintf("apply ConfigMap/%s-sql: %v", deployment, err), Cause: err}
	}
	manifest := RenderFlinkDeploymentManifest(cfg, namespace, deployment, topology, streams)
	if err := d.Applier.Apply(ctx, namespace, "FlinkDeployment", deployment, manifest); err != nil {
		return DeploymentReport{}, &DeployerError{Kind: DeployerErrKube, Message: fmt.Sprintf("apply FlinkDeployment/%s: %v", deployment, err), Cause: err}
	}
	if d.Recorder != nil {
		if err := d.Recorder.RecordTopologyDeployment(ctx, topology.ID.String(), deployment, namespace); err != nil {
			return DeploymentReport{}, &DeployerError{Kind: DeployerErrDB, Cause: err}
		}
	}
	return DeploymentReport{
		Coords: FlinkJobCoords{
			DeploymentName: deployment,
			Namespace:      namespace,
		},
		SQL: sql,
	}, nil
}

// DeleteTopology mirrors deployer::delete_topology — best-effort
// teardown of the ConfigMap + FlinkDeployment owned by this topology.
// Errors from the applier are surfaced so callers can log them, but the
// applier itself is expected to swallow 404s.
func (d *Deployer) DeleteTopology(ctx context.Context, coords FlinkJobCoords) error {
	if d == nil || d.Applier == nil {
		return &DeployerError{Kind: DeployerErrKube, Message: "applier is not wired"}
	}
	if coords.DeploymentName == "" || coords.Namespace == "" {
		return &DeployerError{Kind: DeployerErrKube, Message: "deployment coords are incomplete"}
	}
	if err := d.Applier.Delete(ctx, coords.Namespace, "FlinkDeployment", coords.DeploymentName); err != nil {
		return &DeployerError{Kind: DeployerErrKube, Message: fmt.Sprintf("delete FlinkDeployment/%s: %v", coords.DeploymentName, err), Cause: err}
	}
	if err := d.Applier.Delete(ctx, coords.Namespace, "ConfigMap", coords.DeploymentName+"-sql"); err != nil {
		return &DeployerError{Kind: DeployerErrKube, Message: fmt.Sprintf("delete ConfigMap/%s-sql: %v", coords.DeploymentName, err), Cause: err}
	}
	return nil
}

// RenderSQLConfigMap builds the ConfigMap manifest as a JSON-shaped map.
// Pure helper exposed for unit tests and the kube applier shim.
func RenderSQLConfigMap(namespace, deployment, script string) map[string]any {
	return map[string]any{
		"apiVersion": "v1",
		"kind":       "ConfigMap",
		"metadata": map[string]any{
			"name":      deployment + "-sql",
			"namespace": namespace,
			"labels": map[string]any{
				"app.kubernetes.io/managed-by": FieldManager,
				"openfoundry.io/component":     "flink-sql",
			},
		},
		"data": map[string]any{
			"topology.sql": script,
		},
	}
}

// RenderFlinkDeploymentManifest mirrors deployer::render_flink_deployment_manifest.
// Pure function: builds the FlinkDeployment manifest as a JSON-shaped
// map. Exposed so the REST handler can preview the manifest without
// touching the cluster.
func RenderFlinkDeploymentManifest(cfg FlinkRuntimeConfig, namespace, deployment string, topology *domain.TopologyDefinition, streams []domain.DomainStreamDefinition) map[string]any {
	checkpointDir := strings.TrimRight(cfg.StateBucketURI, "/") + "/checkpoints/" + deployment
	savepointDir := strings.TrimRight(cfg.StateBucketURI, "/") + "/savepoints/" + deployment
	haDir := strings.TrimRight(cfg.StateBucketURI, "/") + "/ha/" + deployment
	checkpointingMode := "AT_LEAST_ONCE"
	if EffectiveExactlyOnce(topology, streams) {
		checkpointingMode = "EXACTLY_ONCE"
	}

	flinkConfig := map[string]any{
		"taskmanager.numberOfTaskSlots":       "2",
		"parallelism.default":                 "4",
		"high-availability.type":              "kubernetes",
		"high-availability.storageDir":        haDir,
		"state.backend.type":                  "rocksdb",
		"state.backend.incremental":           "true",
		"state.checkpoints.dir":               checkpointDir,
		"state.savepoints.dir":                savepointDir,
		"execution.checkpointing.mode":        checkpointingMode,
		"execution.checkpointing.interval":    fmt.Sprintf("%d", topology.CheckpointIntervalMS),
		"execution.checkpointing.timeout":     "600000",
		"metrics.reporter.prom.factory.class": "org.apache.flink.metrics.prometheus.PrometheusReporterFactory",
		"metrics.reporter.prom.port":          "9249",
	}

	return map[string]any{
		"apiVersion": "flink.apache.org/v1beta1",
		"kind":       "FlinkDeployment",
		"metadata": map[string]any{
			"name":      deployment,
			"namespace": namespace,
			"labels": map[string]any{
				"app.kubernetes.io/managed-by": FieldManager,
				"openfoundry.io/topology-id":   topology.ID.String(),
				"openfoundry.io/topology-name": sanitizeLabel(topology.Name),
			},
		},
		"spec": map[string]any{
			"image":              cfg.SQLRunnerImage,
			"flinkVersion":       cfg.FlinkVersion,
			"mode":               "native",
			"serviceAccount":     "flink",
			"flinkConfiguration": flinkConfig,
			"jobManager": map[string]any{
				"replicas": 1,
				"resource": map[string]any{
					"memory": "1024m",
					"cpu":    1,
				},
			},
			"taskManager": map[string]any{
				"resource": map[string]any{
					"memory": "2048m",
					"cpu":    1,
				},
			},
			"podTemplate": map[string]any{
				"spec": map[string]any{
					"containers": []any{
						map[string]any{
							"name": "flink-main-container",
							"volumeMounts": []any{
								map[string]any{
									"name":      "topology-sql",
									"mountPath": "/opt/flink/usrlib/sql",
								},
							},
						},
					},
					"volumes": []any{
						map[string]any{
							"name": "topology-sql",
							"configMap": map[string]any{
								"name": deployment + "-sql",
							},
						},
					},
				},
			},
			"job": map[string]any{
				"jarURI":      "local:///opt/flink/usrlib/sql-runner.jar",
				"args":        []any{"--script", "/opt/flink/usrlib/sql/topology.sql"},
				"parallelism": 2,
				"upgradeMode": "savepoint",
				"state":       "running",
			},
		},
	}
}

// sanitizeLabel mirrors deployer::sanitize_label. Truncates to 63 chars
// (the Kubernetes label-value limit) and replaces any non
// `[A-Za-z0-9-_.]` character with `-`.
func sanitizeLabel(s string) string {
	if len(s) > 63 {
		s = s[:63]
	}
	out := make([]byte, 0, len(s))
	for i := 0; i < len(s); i++ {
		ch := s[i]
		switch {
		case ch >= 'a' && ch <= 'z',
			ch >= 'A' && ch <= 'Z',
			ch >= '0' && ch <= '9',
			ch == '-', ch == '_', ch == '.':
			out = append(out, ch)
		default:
			out = append(out, '-')
		}
	}
	return string(out)
}
